<img src="Figures/top_ML.png" alt="Drawing" style="width: 1000px;"/>

# Introducción a Python    

### CONTENIDOS
1. Introducción a Jupyter Notebooks 
2. Variables y estructuras de datos 
3. Estructuras condicionales 
4. Listas y loops 
5. Diccionarios y tablas de frecuencia
6. Funciones 
7. Funciones avanzadas 
8. Lenguaje orientado a objetos 
9. Introducción a NumPy 
10. Introducción a Pandas
11. Creación de gráficos y visualización

## 1. Introducción a Jupyter Notebooks

<img src="Figures/Jupyter.png" alt="Drawing" style="width: 200px;"/>

Jupyter notebooks es un entorno (IDE) que nos permite intercalar contenido de texto (se llamará cela 'Markdown') con código Python (tipo de cela 'Code') que se puede ejecutar cela por cela como se puede ver en los siguientes ejemplos. 

Los comando básicos para poder seguir este curso y empezar a usar Jupyter Notebooks son:

<img src="Figures/menu.png" alt="Drawing" style="width: 500px;"/>

* `Shift+Enter`: Ejecutar cela y seleccionar siguiente cela
* `Alt+Enter`: Ejecutar cela y insertar una cela nueva
* `ESC`: para entrar en modo 'command'. Una vez en este modo:
    * `H` lista de comandos
    * `A` insertar cela debajo
    * `B` insterar cela arriba
    * `D,D` D dos veces eleimina la cela
    * `Y` cambiar cela a tipo 'Code'
    * `M` cambiar cela a 'Markdown'
    * `Enter`pasar a modo edit

#### Ejecutar la cela de debajo usando alguno de los comandos aprendidos

In [ ]:
3 + 4 + 9

In [ ]:
# Ejemplo de código en Python
# Cambiar los valores de a y b, y observar el resultado al ejecutar la cela 

a=87
b=34
c=a+b

print('El resultado de sumar {} y {} es {}.'.format(a,b,c))

## 2. Variables y estructuras de datos

Las variables en Python se crean cuando se definen por primera vez, es decir, cuando se les asigna un valor por primera vez (utilizando el operador <code> = </code>). Por esta razón, en Python no hace falta declarar las variables. 

Los principales tipos de variables que se pueden encontrar son:
* **Numéricas**
    * **Integer (int):** Números naturales positivos y negativos.
    * **Float:** Números reales.
    * **Complex:** Números complejos. La parte imaginaria se multiplica por `j` para representar la raíz de `-1`. Por ejemplo: <code>a = 2+3j</code>
* **Boleanas**
    Son variables que pueden tomar como valor cierto o falso, representados en Python por `True` y `False` usando mayúsculas.
* **Estructuras de datos (secuencias de datos)**
    Lista ordenada de valores del mismo o distinto tipo. En Python existen:
    * **String:** cadenas de texto
    * **Lista:** conjuntos ordenados de elementos
    * **Tupla:** conjuntos ordenados e inmutables de elementos
    * **Diccionario:** conjuntos de elementos caracterizados por un identificador y un valor    

Existen algunas reglas generales para los nombres de variables:
+ Se usan solo letras, números y "guiones bajos" (_)
+ NO se usan espacios en ningún lugar del nombre
+ El nombre no puede empezar con un número
+ Las letras mayúsculas se tratan como distintas de las minúsculas (es decir, Python distingue entre mayúsculas y minúsculas)

In [ ]:
3a = 10

Cada vez que se crea o modifica una variable, Python interpreta qué tipo de variable es. Cuando se asigna un nuevo valor a un objeto existente (*escritura dinámica*), los valores anteriores de ese objeto se borran de la memoria de la computadora. El primer valor de **a** es el número real 13.0, pero después de una nueva asignación:

In [ ]:
a = 13.0
a = 'b'
print(a)

La función ``type()`` muestra el TIPO del dato que le pasamos como argumento:

In [ ]:
# Declarar una variable a y mirar qué tipo de variable es utilizando "type(nombre_variable)"

a=2
type(a)

In [ ]:
# Probar de declarar otro tipo de variable y mirar qué tipo es

a=2.0
type(a)

### 2.1 Tipos y operadores básicos

Los operadores básicos en Python son:

+ Suma (también para strings, tuplas o listas): <code>a + b</code>
+ Resta: <code>a - b</code>
+ Multiplicación (también para strings, tuplas y listas): <code>a * b</code>
+ División: <code>a / b</code>
+ Division entera: <code>a // b</code>
+ Resto de la división o residuo: <code>a % b</code>
+ Exponencial: <code>a ** b</code>
+ Asignamiento: <code>=</code>, <code>-=</code>, <code>+=</code>,<code>/=</code>,<code>*=</code>, <code>%=</code>, <code>//=</code>, <code>**=</code>
+ Comparaciones booleanas: <code>==</code>, <code>!=</code>, <code><</code>,<code>></code>,<code><=</code>, <code>>=</code>
+ Operaciones booleanas: <code>and</code>, <code>or</code>, <code>not</code>
+ Operaciones para comprobar pertenecencia: <code>in</code>, <code>not in</code>
+ Operaciones para identificar objetos: <code>is</code>, <code>is not</code>
<!-- + Bitwise operators (or, xor, and, complement): <code>|</code>, <code>^</code>, <code>&</code>, <code>~</code> -->
<!-- + Left and right bit shift: <code><<</code>, <code>>></code> -->

### Python como calculadora

Python tiene una notación concisa para la aritmética que se parece mucho a la forma tradicional de escribir operaciones.

In [ ]:
a = 3+2
b= 3.5 * -8
c = 10/6
print(a, b, c, 10./6.)

In [ ]:
# Ejemplo

a = 12
b = 13.3
c = a + b
print (c, type(a), type(b), type(c))

### Módulos 

Algunas funciones matemáticas más complejas no están disponibles en el módulo básico de Python y deben importarse desde un módulo específico. Para importar librerías, solo se debe escribir "import + el nombre de la librería".

<img src="Figures/python_import.jfif" alt="Drawing" style="width: 250px;"/>

En cada caso se deberá leer bien la documentación de las librerías usadas para concocer sus características y saber como implementarlas. Por ejemplo, para math:

https://docs.python.org/3/library/math.html

In [ ]:
import math   # solo se ejecutará si la librería no ha estado importada previamente
print(math.pi + math.sin(100) + math.ceil(2.3))

### Operadores booleanos

In [ ]:
a = 4
b = 40
(a>2) and (b>30)

In [ ]:
(a>2) or (b>100)

In [ ]:
not(a>2)

### 2.2 Estructuras de datos

### Strings o cadenas de texto

El texto en Python se trata como una lista de carcateres. De esta manera, la metodología que aplica al tratamiento de listas, puede ser aplicada a las cadenas de texto.

In [ ]:
a = 'python'
type(a)

In [ ]:
print("Hello" )

In [ ]:
print("Esto es 'un ejemplo' de como usar comillas dentro de un texto que se escribe entre comillas dobles")
print('Aquí tenemos "otro ejemplo" pero ahora al revés')

Para juntar cadenas de carácteres se puede usar el operador '+':

In [ ]:
a = 'He'
b = 'llo'
c = a+b+'!'
print(c)

Para acceder a solo una parte de una lista o, en este caso, de una *string*, se usa la técnica denominada **slicing**. El formato a seguir para coger solo una parte de la string es el siguiente:

# <center>nombre_variable[indice_primer_elemento:indice_final:step]</center>

Donde el funcionamiento de los índices se pueden ver en la siguiente imagen

![Slicing1](Figures/slicing1.png)


In [ ]:
#Ejemplos
a = 'Python'
print(a[:], a[0], a[2:], a[:3], a[2:4], a[::2], a[1::2])

#### Ejercicio
Con una línia de codigo, conseguir los siguientes sets de caracteres:
- Primera letra.
- De la letra 2 a la 4
- Úlimas 3 letras
- Solo las letras en posición par, des de la número 4 al final
- Todo el texto al revés


In [ ]:
texto = "Texto de prueba para hacer el ejercicio"
primera_letra = print(texto[0])

En el caso de los *strings* hay métodos en Python que vienen por defecto que nos pueden ser útiles. Por ejemplo las funciones siguientes:
- len: devuelve la longitud el texto.
- str: convierte eun número en string.
- reverse: devuelve el texto al revés.

Para usar una fucnión se pondrá nombre_de_la_función(nombre_variable)

También podemos usar métodos propios para strings:
- lower: convierte todo el texto en mínusculas.
- upper: convierte todo el texto en mayúsculas.
- capitalize: pone en mayúsculas la primera letra del string.
- split(delimitador): devuelve una lista de substrings separados por el delimitador indicado (ver ejemplos más adelante)

Para usar un método se pondrá nombre_variable.nombre_del_métdodo


In [ ]:
texto='vamos a hacer pruebas con las funciones y métodos de Strings!'
print(len(texto))
print(texto.lower())
print(texto.split(' '))
# En una línia, devuelve la 4a palabra del texto todo en mayúsculas!


### Lists

Las listas son un tipo de Python que sirven para agrupar objetos de otros tipos (incluso listas!). Por ejemplo una lista de números, de floats, de strings, de otras listas, etc. 

Las listas, como los strings, pueden usar la metodología de *slicing*


In [ ]:
l=[]
type(l)

In [ ]:
# Crear lista
mi_lista = [0,5,'6',34,'Hola']
mi_lista

In [ ]:
# Acceder a un elemento de la lista
a = mi_lista[2]
a

In [ ]:
# Canviar un elemento de una lista
mi_lista[-1]='Adiós'
mi_lista

In [ ]:
# Podemos tener lista de listas!
mi_lista=[45,[3,5,6],'palabra',34.6,['a','b','c']]
print(mi_lista)
print(mi_lista[1])
print(mi_lista[1][2])
# acceder a la letra c!

Para las listas también hay funciones y métodos en Python que vienen por defecto que nos pueden ser útiles. Por ejemplo las funciones siguientes:
- len: devuelve la longitud de la lista.
- list: crear lista

Para usar una fucnión se pondrá nombre_de_la_función(nombre_variable)

También podemos usar métodos propios para listas:
- append: añadir elemento al final de la lista.
- insert: añdir un elemento en una posición particular.
- pop: quitar el último elemento de la lista.
- remove: quitar un elemento en particular.
- sort: ordenar la lista

Para usar un método se pondrá nombre_variable.nombre_del_métdodo

In [ ]:
mi_lista=['primero','hola','prueba','adiós','zoológico','máster']
print(mi_lista)
mi_lista.append('último')
print(mi_lista)
mi_lista.insert(1,'segundo')
print(mi_lista)
mi_lista.pop()
print(mi_lista)
mi_lista.remove('prueba')
print(mi_lista)
mi_lista.sort()
print(mi_lista)

### Dictionaries

Un diccionario es otra manera de agrupar datos donde se puede acceder a un *element* a través de su *key*. Por ejemplo:

In [ ]:
esta_instalacion = {
    'Potencia': 10,
    'Tipo': 'Solar',
    'Año_construccion':2012
}

In [ ]:
esta_instalacion['Tipo']

Otro ejemplo podría ser:

In [ ]:
dict = {"Instalacion_1": "Solar", "Instalacion_2": "Eólica", "Instalacion_3": "Eólica", "Instalacion_4": "Hidráulica", "Instalacion_5": "Solar",}
print(dict)

Para acceder a algun elemento, modificar o añadir, se podrá hacer de forma parecida a las listas:

In [ ]:
#acceder
dict["Instalacion_1"]

In [ ]:
# modificar
print(dict)
dict["Instalacion_1"]="Eólica"
print(dict)

In [ ]:
print(dict)
dict["Instalacion_6"]="Solar"
print(dict)

Los diccionarios pueden contener todos los tipos (int, floats, listas, otros diccionarios, etc.)

In [ ]:
dict = {"Instalacion_1": ["Solar", '10 MW'], "Instalacion_2": ["Eólica", '15 MW'], "Instalacion_3": ["Solar", '18 MW'], "Instalacion_4": ["Hidráulica", '12 MW'], "Instalacion_5": ["Solar", '23 MW'], "Instalacion_6": ["Eólica", '8 MW'],"Instalacion_7": ["Solar", '11 MW'], "Instalacion_8": ["Hidráulica", '11 MW']}
dict

In [ ]:
dict['Instalacion_3'][1]

In [ ]:
#Suma 3 MW a la instalación 4 y modifica el diccionario!

### Tuples

Tuples son listas **que no permiten ser modificadas!**. 

Su sintaxis es igual que en las listas, però para escribir una tupla se usan paréntesis y no corchetes.

In [ ]:
misupertupla = ('Solar', 'Eólica', 'Hidráulica', 'Gas', 'Nuclear', 'NoQueremosMásCarbón')
print(type(misupertupla), misupertupla[4:])

In [ ]:
misupertupla[0]='PV'

## 3. Estructuras condicionales (if)

Las estructuras de control condicionales permiten ejecutar una parte del código o otro en función de la evaluación de una o varis condiciones booleanas de SI o NO (**<code>TRUE</code>** o **<code>FALSE</code>**).

En Python, las estructuras de control condicionales se definen mediante las palabras **<code>if</code>**, **<code>elif</code>** y **<code>else</code>**.

- **<code>if CONDICION:</code>** si se cumple la expresión condicional se ejecuta el bloque de código a continuación.
- **<code>elif CONDICION:</code>** de lo contrario, si se cumple esta expresión condicional se ejecuta este otro bloque de código.
- **<code>else:</code>** de lo contrario, se ejecuta este bloque de código a continuación sin evaluar ninguna condición.


La condición es normalmente evaluada mediante los siguientes operadores relacionales <code> <, <=, ==, >=, >, != </code>.

La parte de código dentro de un ``if`` no pueden estar vacía, pero si por alguna razón se debe crear una declaración if sin contenido, se puede utilizar la declaración ``pass`` para evitar de se produzca un error.

In [ ]:
# Ejemplo 

a = 33
b = 200

if b > a:
  pass

In [ ]:
#Ejemplo

celsius = 35
fahrenheit = 9.0 /5.0 * celsius + 32

print("La temperatura en Fahrenheit es", fahrenheit)

if fahrenheit > 90:
    print("Hace calor")
elif fahrenheit < 30:
    print("Hace frío")
else: print("No hace ni frío ni calor")

#### Ejercicio

Crear un código para identificar si la variable **<code>numero</code>** es un numero positivo, negativo, o de valor 0. Guardar el resultado en una variable llamada **<code>signo</code>**

In [ ]:
numero = 2


Tambiém se puede expresar en una sola linea la estructura condicional de la siguiente manera, dependindo de si se trata de una, dos o más condiciones:

In [ ]:
#Ejemplo

a = 330
b = 330

# con una sola condición:
if a > b: print("a es mayor que b")
    
#con dos condiciones:
print("A") if a > b else print("B")

#con tres condiciones:
print("A") if a > b else print("=") if a == b else print("B")

In [ ]:
# Ejercicio: poner en una sola linea el código para determinar el signo de un número 
numero = 2


## 4. Listas y loops

### 4.1 Estructuras de control iterativas (for & while)

Las estructuras de control iterativas (también llamadas bucles o loops), permiten ejecutar un mismo código de manera repetida mientras se cumpla una condición.

Existen dos tipos de estructuras de control iterativas:

* For
* While

La sentencia ``for`` de Python itera sobre los elementos de cualquier secuencia (una lista o una cadena de caracteres por ejemplo), en el orden que aparecen en la secuencia.

El ciclo ``while`` permite realizar múltiples iteraciones basándose en el resultado de una expresión lógica que puede tener como resultado un valor ``True`` o ``False``.


In [ ]:
#Ejemplo de for

for i in range(0,10,1):
    print(i)

In [ ]:
#Ejemplo de while

i=0
while i < 10 :
    print(i)
    i= i+1

Además, los bucles con ``for`` o ``while`` pueden combinarse con estructuras condicionales ``if`` dando lugar a estructuras más complejas.

Con la instrucción ``break`` podemos detener el bucle antes de que haya pasado por todos los elementos.

Con la instrucción ``continue`` podemos detener la iteración actual del bucle (tanto en ``for`` como en ``while``) y continuar con la siguiente.

In [ ]:
# Ejemplo

numeros = [-5, 3, 2, -1, 9, 6]
total = 0

for n in numeros:
    if n >= 0:
        total += n

total2 = 0
for n in numeros:
    if n < 0:
        continue
    total2 += n

print(total, total2)

In [ ]:
#Ejercicio: calcular el valor medio de una lista con un for

lista=[1,2,3,4]


In [ ]:
#Ejercicio: rehacer el código anterior para calcular el valor medio con un while en lugar de un for


### 4.2 Lists (o dictionary) comprehensions

Las comprensiones de listas son una forma de ajustar un bucle ``for``, una declaración ``if`` y una asignación, todo en una sola línea.

Una lista de comprensión consta de las siguientes partes:

+ Una secuencia de entrada
+ Una variable que representa miembros de la secuencia de entrada
+ Una expresión opcional
+ Una expresión de salida que produce elementos de la lista de salida a partir de miembros de la secuencia de entrada que satisfacen el predicado

In [ ]:
# Ejemplo
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = []
for i in num:
     if i > 0:
        squared.append(i**2)
print(type(squared), squared)

# Este for se puede reescribir de la siguiente manera:
# En una iteración, primero comprueba si el if es TRUE y luego realiza el cálculo
squared = [ x**2 for x in num if x > 0] 
print(type(squared), squared)

In [ ]:
#Ejercicio: calcular el valor medio de una lista con un for de una sola linea

lista=[1,2,3,4]


Esta forma compacta de expresar los bucles también se puede utilizar para iterar dentro de listas. A contnuación se muestran diferentes maneras de juntar los valores de dos listas usando zip o usando fors anidados.

In [ ]:
lista1 = ['a','b','c','d']
lista2 = [1,2,3,4]

# crear una lista con fors anidados:
print ([(i,j) for i in lista1 for j in lista2])

# crear una lista usando zip:
print ([(i,j) for i,j in zip(lista1,lista2)])

# crear un diccionario usando zip:
print ( {i:j for i,j in zip(lista1,lista2)} )

## 5. Diccionarios y tablas de frecuencia

Los diccionarios son muy útiles para contar elementos en una lista, y crear una distribución de estos mediante una tabla de frecuencia. A continuación se puede observar un ejemplo donde se quiere saber qué número del 1 al 9 aparece más veces en una lista desordenada de números.

In [ ]:
mi_lista = [2,3,5,2,5,6,7,8,1,2,3,4,5,2,8,9,5,2,4,6,9,1,2,1,3,6,7,4,3,1,2,9,7,5,4,3,1,2,9,5,4,2,7,9,5,3,4,6,7,3,3,4,5]

In [ ]:
# Creamos un diccionario vacío 
freq = {} 
for numero in mi_lista: 
    if (numero in freq): 
        freq[numero] += 1
    else: 
        freq[numero] = 1

for key, value in freq.items(): 
    print ("% d : % d"%(key, value))

De una lista de instalaciones, obtenemos estas tipologías de plantas de generación. ¿Qué tecnología tiene más plantas instaladas? 

¿Detectas algún problema con los nombres? ¿Cómo lo solucionarías?

In [ ]:
instalaciones = ['Eólica', 'Solar', 'Hidráulica', 'Solar', 'Eólica', 'Hidráulica', 'Gas Natural', 'Solar', 'Eólica', 'Solar', 'Nuclear', 'Solar', 'Hidráulica', 'Hidráulica', 'Solar', 'Carbón', 'Eólica', 'Eólica', 'Solar', 'Solar', 'Solar', 'Hidráulica', 'Gas Natural', 'Hidráulica', 'Nuclear', 'Eólica', 'Eólica', 'Solar', 'Eólica', 'Gas Natural', 'Hidráulica', 'hidráulica', 'Solar', 'Eólica', 'Hidráulica', 'Solar', 'Eolica', 'Solar','eolica', 'Eólica', 'Solar', 'Solar', 'Hidráulica', 'Gas Natural', 'Solar', 'Gas Natural', 'Nuclear', 'eólica', 'Hidráulica', 'Hidráulica', 'Solar']
instalaciones

## 6. Funciones

Una función es un bloque de código con un nombre asociado, que recibe cero o más argumentos como entrada, sigue una secuencia de sentencias donde se ejecutan las operaciones deseadas y devuelve un valor y/o realiza una tarea. Este bloque de código puede ser llamado varias veces cuando se necesite.

En Python existen una serie de funciones integradas por defecto al lenguaje, pero también se pueden crear funciones definidas por el usuario para utilizarlas en sus propios programas. 

La lista completa de funciones integradas por defecto puede consultarse en:
www.w3schools.com/python/python_ref_functions.asp

In [ ]:
#Ejemplo de función integrada en Python

# La función input() nos permite asignar a una variable un valor ingresado por el usuario

a = input()

print('La variable introducida ha sido ' + a)

Para crear una función en python, se debe utilizar la palabra ``def`` para definir la función. La sintaxis para una definición de función en Python es:

<code>*def* nombre_funcion(lista_de_parametros_entrada):
    sentencias_funcion
    *return* [expresion_a_devolver]</code>

Donde:
* nombre_funcion: es el nombre de la función.
* lista_de_parametros_entrada: es la lista de parámetros que puede recibir una función, separados por coma.
* sentencias_funcion: es el bloque de sentencias en código fuente Python que realizar cierta operación dada.
* expresion_a_devolver: es la expresión o variable que devuelve la sentencia return.

In [ ]:
# Ejemplo: función para poner a 0 el primer valor de una lista

def change_first_element(list):
    list[0]=0 #it returns none!

numbers=[1,2,3,4]
change_first_element(numbers)
print(numbers)

#### Ejemplo: máximo común denominador

El común denominador de dos enteros positivos $a$ y $b$ es el mayor divisor común entre $a$ y $b$.  El algortimo Euclideo es un método iterativo para calcular el máximo común denominador de dos enteros. En pseudocódigo seria:

+ Si $a<b$, intercanviar $a$ y $b$.
+ Dividir $a$ entre $b$ y obtener el residuo, $r$. 
+ Si $r=0$, el valor $b$ es el MCD de $a$ y $b$.
+ Si $r \neq 0$, iterar de nuevo reemplazando $a$ por $b$ y reemplazando $b$ por $r$. 

In [ ]:
def mcd(a,b):
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

mcd(100,16)

#### Ejercicio: crear una función para calcular el factorial de un entero

El factorial de un entero no negativo 𝑛 es el producto de todos enteros positivos menores o iguales a  𝑛.


In [ ]:
# Poner tu código aquí:


### 6.1 Argumentos y parámetros de funciones

Al definir una función, los valores que se reciben se denominan parámetros, mientras que durante la llamada, los valores que se envían se denominan argumentos.

Por defecto, se debe llamar a una función con el número correcto de argumentos. Si la función espera 2 argumentos, debe llamar a la función con 2 argumentos.

#### Argumentos por posición

Cuando se envian argumentos a una función, por defecto estos se reciben por orden en los parámetros definidos, lo que se conoce como argumentos por posición. En el ejemplo anterior, el argumento 100 es la posición 0, que se corresponde al parámetro de la función a, mientras que el argumento 16 es la posición 1, que se corresponde al parámetro de la función b.

#### Argumentos por nombre

Sin embargo, es posible saltarse el orden de los parámetros si se indica durante la llamada que valor tiene cada parámetro a partir de su nombre.

In [ ]:
mcd(b=100,a=16)

#### Parámetros por defecto

Para evitar mensajes de error cuando no se llama a una función con los argumentos necesarios, se pueden definir parámetros por defecto dentro de una función.

In [ ]:
def mcd(a=None,b=None):
    if a == None or b == None:
        print ("Error, debes enviar dos números a la función")
        return
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

mcd()

#### Parámetros indeterminados ``*args`` y ``**kwargs``

En algunas ocasiones, no se sabe con anterioridad cuantos elementos es necesario enviar a una función.

Si no se sabe cuántos argumentos por posición se pasarán a una función, se agrega un * antes del nombre del parámetro en la definición de la función. Por convenio, se suele utilizar el nombre de parámetro ``*args``. De esta forma, la función recibirá una tupla de argumentos y podrá acceder a los elementos en consecuencia.

De forma similar, si no se sabe cuántos argumentos por nombre se pasarán a una función, se agregan dos ** antes del nombre del parámetro en la definición de la función. Por convenio, se suele utilizar el nombre de parámetro ``**kwargs``. De esta forma, la función recibirá un diccionario de argumentos y podrá acceder a los elementos en consecuencia.

In [ ]:
def super_funcion(*args,**kwargs):
    total = 0
    for arg in args:
        total += arg
    print ("suma => ", total)
    for kwarg in kwargs:
        print (kwarg, "=>", kwargs[kwarg])

super_funcion(50, -1, 1.56, 10, 20, 300, nombre="Pedro", edad=38)

## 7. Funciones avanzadas

En este apartado se verá como en alguna ocasiones, hay funciones simples que solo queremos usar una vez. Este uso no parece muy apropiado para las funciones de las que hemos hablado hasta ahora... Vamos a ver primero ejemplos dónde este tipo de funciones nos pueden ser de utilidad. Se trata de las funciones <code>map, filter y reduce<code>.

### 7.1 Map

Con un "map" podemos aplicar una función a toda una lista de inputs. Por ejemplo si queremos coger una lista de números y aplicar el cuadrado a todos sus elementos podríamos hacer:


In [ ]:
items = [1, 2, 3, 4, 5]
squared = []
for i in items:
    squared.append(i**2)
print(items)
print(squared)

Usando ``Map`` , tendríamos:

In [ ]:
def square_item (x):
    return x**2
squared = list(map(square_item, items))
print(items)
print(squared)

Como se puede ver ``map`` es una función que acepta una función como argumento

``map(funcion a aplicar, lista de entrada)``



### 7.2 Filter

La función filter de una manera similar, devuelve una lista con aquellso elementos de una lista de entrada que cumple una cierta codició (filtra la lista de etrada con una condición determinada). Véamos como ejemplo, si queremos quitar los elementos negativos de una lista y quedarnos solo con los positivos y que sean pares

In [ ]:
items=[2,5,-6,4,-1,2,7,-8,3,4,7,0,-1,-4,0,4,6,-8,0,1]
positivos=[]
for item in items:
    if item>=0:
        if item % 2 == 0:
            positivos.append(item) 
print(items)
print(positivos)

In [ ]:
items=[2,5,-6,4,-1,2,7,-8,3,4,7,0,-1,-4,0,4,6,-8,0,1]
def es_positivo(x):
    return x>=0 and x%2==0
positivos=list(filter(es_positivo, items))
print(items)
print(positivos)

Se puede observar como su sitaxis es muy parecisa a la funció ``map``, pero usando ahora ``filter``

``filter(funcion a aplicar, lista de entrada)``

### 7.3 Reduce

La función ``Reduce`` será muy útil cuando se quiera hacer un cálculo sobre una lista y obtener el resultado. Por ejemplo si queremos obtener el resultado de multiplicar los números de una lista haríamos:

In [ ]:
producto = 1
lista = [1,5,7,3,4,9,1,4,7,5,3,9,8,7,2]
for num in lista:
    producto = producto * num
print(producto)

Usando ``reduce``:

In [ ]:
from functools import reduce
lista = [1,5,7,3,4,9,1,4,7,5,3,9,8,7,2]
def multiplicar(x,y):
    return x*y
producto = reduce(multiplicar, lista)
print(producto)

Vemos que para usar ``reduce`` se tiene que importar de ``functools`` y que tiene una sintaxis idética a los ateriores casos:
``reduce(funcion a aplicar, lista de entrada)``

### 7.4 Funciones Lambda 

En todos estos casos hemos visto como consegumos comprimir el código, pero siempre tenemos que definir fuciones, normalmente muy simple, solo para ser usadas una vez. Para estas ocasiones, existen las ``funciones anónimas`` o ``lambda functions``. Estas se pueden definir y usar en la misma línia, y nos ahorramos un montón de tiempo!!

Su sintaxis es:

``lambda entradas: salida``

Vamos a volver a escribir las mismas funciones que antes, perpo usado estas funciones para verlo más claro con ejemplos:

In [ ]:
items = [1, 2, 3, 4, 5]
squared = list(map(lambda x:x**2, items))
print(items)
print(squared)

In [ ]:
items=[2,5,-6,4,-1,2,7,-8,3,4,7,0,-1,-4,0,4,6,-8,0,1]
positivos=list(filter(lambda x: x>=0 and x%2==0, items))
print(items)
print(positivos)

In [ ]:
from functools import reduce
lista = [1,5,7,3,4,9,1,4,7,5,3,9,8,7,2]
producto = reduce(lambda x,y: x*y, lista)
print(producto)

#### Ejercicio

Dada una lista de nombres de clientes, devolver la lista con todos los nombres con la primera letra en mayúsculas

In [ ]:
nombres = ['jordi', 'arnau', 'laia', 'pedro', 'juan', 'pau', 'noemi', 'arya', 'marti', 'lyanna', 'miquel', 'joana', 'ricard', 'daenerys', 'alexandra']
# vuestro código aquí!

#### Otros usos

Hay más ocasiones donde una función pide otra función como argumeto, dode usar una función lambda será muy útil. Por ejemplo, si miramos la documentación de la función ``sorted`` que viene en Python por defecto:

https://docs.python.org/3/library/functions.html#sorted

In [ ]:
desordenado = ['Martí','Anna','daniel','Pau','berta','Roberto','Xavier', 'alexandra']
sorted(desordenado)

Si especificamos una función en el apartado ``key`` opcional:

In [ ]:
sorted(desordenado, key=lambda x: x.lower())

## 8. Lenguaje orientado a objetos

Python es un lenguaje de programación orientado a objetos. Casi todo en Python es un objeto, con sus propiedades y métodos. Una clase es como un constructor de objetos, o un "plano" para crear objetos.

Entre los conceptos asociados a una clase destacan los siguientes:

+ Clase: una especie de "plantilla" en la que se definen los atributos y métodos predeterminados de un tipo de objeto.

+ Objeto: instancia de una clase.

+ Método: son funciones asociadas a un objeto o a una clase de objetos. Se podria decir que es lo que el objeto puede hacer.

+ Atributos: son variables asociadas a un objeto o a una clase de objetos. Se podria decir que son las características que tiene una clase. Hay atributos de instancia para inicializar los atributos mínimos de la clase cuando se crea la clase, y atributos de clase para complmentar la descripción del objeto.

<img src="Figures/clases.png" alt="Drawing" style="width: 600px;"/>

En Python se define una clase con la palabra reservada ``class``.

El parámetro ``self`` es una referencia a la instancia actual de la clase y se usa para acceder a las variables que pertenecen a la clase.

Una vez creado un objeto, se pueden referenciar a sus atributos o invocar a un método por medio del operador ``.``.

In [ ]:
#Ejemplo: Clase de un rectangulo

class Rectangulo:
    """Clase de un rectangulo"""
    # lista de objetos
    x = 0
    y = 0
    
    # lista de métodos
    def area(self):
        return self.x * self.y

# Crear el objeto 'a' haciendo una instancia a la clase Rectangulo    
a = Rectangulo()
# Inicializar x, y 
a.x = 10
a.y = 4
# Calcular el area del rectangulo
a.area()

### 8.1 Atributos de instancia ``__init__()``

Todas las clases tienen una función llamada ``__init __ ()``, que se ejecuta cuando se inicia la clase y sirve para inicializar los atributos de instancia. Los argumentos que se utilizan en la definición de ``__init__()`` corresponden a los parámetros que se deben ingresar al instanciar un objeto.

In [ ]:
#Ejemplo: clase de un rectangulo

class Rectangulo:
    """Clase de un rectangulo"""
    
    # atributos de instancia
    def __init__(self,x,y):
        self.x = x
        self.y = y

    # lista de métodos
    def area(self):
        return self.x * self.y
    
a = Rectangulo(10,4)

a.area()

#### Ejercicio

Añadir a la clase anterior un método para calcular el perímetro del rectángulo y otro método para cambiar el tamaño del rectángulo mediante un factor de escala.

In [ ]:
# Añadir código aquí


### 8.2 Herencia

En programación orientada a objetos, la herencia es la capacidad de reutilizar una clase extendiendo su funcionalidad. Una clase que hereda de otra puede añadir nuevos atributos, ocultarlos, añadir nuevos métodos o redefinirlos.

#### Ejemplo

En el siguiente ejemplo, la clase alumno hereda de la clase persona para complementar los atributos que puede tener una persona.

In [ ]:
class Persona:
    def __init__(self, nombre, apellido):
        self.nombre = nombre
        self.apellido = apellido

    def imprimirnombre(self):
        print(self.nombre, self.apellido)

class Estudiante(Persona):
    nota = 0
    clase = 'Clase sin introducir'

    def imprimirnota(self):
        print(self.nota)

        
x = Persona("Johann Sebastian", "Mastropiero ")
x.imprimirnombre()

y = Estudiante("Pedro", "Sanchez")
y.nota = 4.99
y.imprimirnombre()
y.imprimirnota()

## 9. Introducción a NumPy

NumPy es una librería de Python utilizada para trabajar con matrices. También tiene funciones para trabajar con álgebra lineal o transformadas de Fourier. NumPy significa Numerical Python.

<img src="Figures/NumPy_logo.png" alt="Drawing" style="width: 400px;"/>

Como hemos visto, en Python ya tenemos listas que sirven al propósito de trabajar con matrices, pero su procesamiento es lento. NumPy tiene como objetivo proporcionar un objeto de matriz que sea hasta 50 veces más rápido que las listas tradicionales de Python, algo importante para la ciencia de datos donde la velocidad y los recursos son muy importantes. Además, se diferencian de las listas en que las matrices de NumPy no se pueden cambiar de tamaño, todos los elementos deben ser del mismo tipo, y se permiten ls operaciones entre matrices.

El código fuente de NumPy se encuentra en este repositorio de github https://github.com/numpy/numpy

El objeto de matriz en NumPy se llama ``ndarray``, proporciona muchas funciones de soporte que hacen que trabajar con ``ndarray`` sea muy fácil. Se puede crear una ndarray  con NumPy con la función ``.array()``

In [ ]:
#Ejemplo
import numpy as np #normalmente se suele importar como np

arr = np.array([1, 2, 3, 4, 5])

print(arr)

print(type(arr))

Para crear un ndarray, podemos utilizar una lista, tupla o cualquier objeto tipo matriz al método ``array()``, y se convertirá en un ``ndarray``. Se pueden crear ``ndarrays`` de diferentes dimensiones. Los 0-D son solo números, los 1-D son vectores, los 2-D son matrices, etc. 

Los principales atributos que se pueden utlizar con ``ndarrays`` son:

| Atributo  | Descripción | 
|-----------|-------------|
| ndim      |número de dimensiones de la matriz| 
| shape     |número de elementos de cada dimensión de la matriz|
| size      |número de elementos total de la matriz|
| dtype     |tipo de los elementos de la matriz|
| astype    |para cambiar el tipo de los elementos de la matriz|

In [ ]:
#Ejemplo

a = np.array(42)
b = np.array([1, 2, 3, 4, 5])
c = np.array([[1, 2, 3], [4, 5, 6]])
d = np.array([[[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]]])

print(a.ndim)
print(b.ndim)
print(c.ndim)
print(d.ndim)
print(d.shape)
print(d.size)
print(d.dtype)

e = d.astype(float)

print(e)
print(e.dtype)

### 9.1 Acceder a valores de un ndarray

Se puede acceder a un elemento de la matriz haciendo referencia a su número de índice. Los índices en las matrices NumPy comienzan con 0, lo que significa que el primer elemento tiene índice 0, y el segundo tiene índice 1, etc.

Para acceder a elementos de matrices 2-D podemos usar enteros separados por comas que representan la dimensión y el índice del elemento. Para matrices 3-D o superior, se deben añadir tantos enteros separados por comas como dimensiones tenga la matriz.

In [ ]:
#Ejemplo

arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print(arr[0, 1])

También se puede utilizar el slicing en ``ndarrays``. En este caso: ``array[indice_primer_elemento:indice_final:step]``

In [ ]:
#Ejercicio

arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])

# devuelve el indice 2 de las dos dimensiones:

# devuelve el intervalo entre el índice 1 y 4 (no incluido) de las dos dimensiones:

# devuelve todos los elementos de posiciones pares:


<img src="Figures/numpy_indexing.png" alt="Drawing" style="width: 200px;"/>

In [ ]:
#Ejercicio: seleccionar los elementos de cada uno de los conjuntos de elementos de la figura anterior:

arr = np.array([[0,1,2,3,4,5],[10,11,12,13,14,15],[20,21,22,23,24,25],[30,31,32,33,34,35],[40,41,42,43,44,45],[50,51,52,53,54,55]])
arr

# Nueva array con selección elementos rojos:
arr2[]
# Nueva array con selección elementos naranjas:
arr3[]
# Nueva array con selección elementos verdes:
arr4[]
# Nueva array con selección elementos azules:
arr5[]

### 9.2 Importar arrays de archivos .csv

Como los archivos ** Comma-separated values (CSV) ** son un tipop de archivo .TXT, se puede importar un archivo .CSV con la función ``np.genfromtxt(ubicacion_archivo)``

Los parametros opcionales de **genfromtxt** pueden consultarse en: https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html

| Función   | Tipo | Descripción | 
|-----------|----|---------|
|skip_header | opcional |número de líneas a evitar al inicio del fichero|
|filling_values | opcional |valores a utilizar cunado no haya valores|
|delimiter | opcional |string utilizado para separar valores (en CSV es ',')|
|usecols | opcional | seleciona que columnas importar, siendo 0 la primera (e.g. ``usecols = (1, 4, 5)``)|


#### Ejercicio: 

Observar el archivo **Data/block_13_diario_reducido.csv** y ver como se ha importado a NumPy.

Sustituir valores nan por ``99999`` y evitar importar la primera fila y las dos primeras columnas.

In [ ]:
data = np.genfromtxt('Data/block_13_diario_reducido.csv',delimiter=',')
print (data)

# Sustituir valores nan por 99999

# Evitar importar la primera fila y las dos primeras columnas


### 9.3 Funciones básicas con NumPy

Aquí podeis ver una lista de las funciones más interesantes a mirar en la documentación, con algun ejemplo:

| Función   | Descripción | 
|-----------|-------------|
| array_name.mean()    |Media| 
| array_name.min()     |Valor mínimo|
| array_name.max()     |Valor máximo|
| array_name.sum()     |Suma de valores|
| array_name.std()     |Desviación standard|
| array_name.var()     |Varianza|
| array_name.reshape(nueva_dimension/es) | Permite cambiar las dimensiones|

Algunos ejemplos de Methods asociados a NumPy:

| Method   | Descripción | 
|-----------|-------------|
| np.amin(array_name)     | valor mínimo|
| np.amax(array_name)     | valor máximo|
| np.argmax(array_name)          | índice del valor máximo|
| np.argmin(array_name)          | índice del valor mínimo|
| np.isnan(array_name)           | identificador de valores Nan|

Para una matriz multidimensional, es posible hacer el cálculo a lo largo de una sola dimensión, pasando el parámetro ``axis`` para indicar el eje. Por ejemplo, en 2-D, axis=0 es la columna  y axis=1 es la fila.

#### Ejercicio

Practicad con alguno de estas funciones o Methods a partir del array ``data``.

In [ ]:
# vuestro código aquí!

# Encontrar valor y posición del valor máximo

# Encontrar valor y posición del valor mínimo de la segunda columna

# Encontrar valor medio de cada fila

# Encontrar posicion de los valores Nan


### 9.4 Filtrado de datos con una matriz booleana

Comparaciones booleanas se pueden usar para comparar elementos en matrices de igual tamaño.

La función ``where`` crea una nueva array a partir de comparar dos arrays, siguiendo la siguiente syntaxis: ``where(bool_array, true_array, false_array)``

In [ ]:
# Ejemplo: evitar error al dividir por cero

a = np.array([1, 3, 0], float) 
np.where(a != 0, 1/a, 0) 

Esto se puede usar para crear una máscara, extrayendo los índices de una matriz que satisfacen una condición dada.

In [ ]:
arr = np.array([10,8,30,40])
print (arr)
mask = arr < 9 # array booleana con los elementos menores a 9
mask

#### Ejercicio

Reescribir el siguiente código para obtener un array donde los valores inferiores a 9 se les imponga el valor 10 utilizando el array mak anterior y la función ``where``.

In [ ]:
print ('Resetting all values below 9 to 10...')
print (arr < 9)
arr[arr < 9] = 10
print (arr)

### Para más información sobre NumPy

* http://numpy.scipy.org
* http://scipy.org/Tentative_NumPy_Tutorial
* https://numpy.org/learn/

## 10. Introducción a Pandas

Padas es de las librerías más importantes en Python para Data Science!!! Básicamente nso permitirá tener ``Datasets`` abiertos como si se trataran de una hoja de cálculo en el excel, es decir, en forma matricial con dos dimensiones y con columnas e idices indexados.

<img src="Figures/Pandas.jpg" alt="Drawing" style="width: 800px;"/>


Pandas sirve para:
* Operar en todo el dataset o a una fila o columna (vectores) con alta eficiencia computacional
* Ayuda a la limpieza de datos y solucionar datos inexistentes
* Acceder a subsets de datos
* Añadir o eliminar columnas con nuevas características.
* Agrupar datos por características.
* Muy eficiente juntando datos de distintas fuentes.
* Trabajar con series temporales.

### 10.1 Leer datos de un csv

La mayoría de las veces cargaremos los archivos a través de un archivo .csv. A continuación podeis ver el código básico que se usará casi siempre:

In [1]:
import pandas as pd #importamos Pandas!
londres = pd.read_csv('Data/block_13_diario.csv')
londres.head(5)

LCLid         day  energy_median  energy_mean  energy_max  \
0  MAC000113  2011-12-14         1.0910     1.065250       1.674   
1  MAC000113  2011-12-15         0.3710     0.546521       1.434   
2  MAC000113  2011-12-16         0.3205     0.534708       2.034   
3  MAC000113  2011-12-17         0.2570     0.421500       1.068   
4  MAC000113  2011-12-18         0.3985     0.565333       2.218   

   energy_count  energy_std  energy_sum  energy_min  
0            28    0.305397      29.827       0.262  
1            48    0.410120      26.233       0.119  
2            48    0.464431      25.666       0.129  
3            48    0.274387      20.232       0.115  
4            48    0.450712      27.136       0.134

### 10.2 Inspección datos

Si queremos saber qué dimensiones tiene este DataFrame, podemos mirar su 'forma'

In [2]:
londres.shape

(32992, 9)

Hay 32992 entradas y 9 características para cada entrada.

Puede ser muy útil ver las primeras o últimas entradas del dataset para entender mejor estos datos:

In [4]:
londres.head(10)

LCLid         day  energy_median  energy_mean  energy_max  \
0  MAC000113  2011-12-14         1.0910     1.065250       1.674   
1  MAC000113  2011-12-15         0.3710     0.546521       1.434   
2  MAC000113  2011-12-16         0.3205     0.534708       2.034   
3  MAC000113  2011-12-17         0.2570     0.421500       1.068   
4  MAC000113  2011-12-18         0.3985     0.565333       2.218   
5  MAC000113  2011-12-19         0.6980     0.746479       2.204   
6  MAC000113  2011-12-20         0.2485     0.436354       1.551   
7  MAC000113  2011-12-21         0.2050     0.426542       1.818   
8  MAC000113  2011-12-22         0.1830     0.181854       0.259   
9  MAC000113  2011-12-23         0.1860     0.185646       0.268   

   energy_count  energy_std  energy_sum  energy_min  
0            28    0.305397      29.827       0.262  
1            48    0.410120      26.233       0.119  
2            48    0.464431      25.666       0.129  
3            48    0.274387      20.232       0.115  
4            48    0.450712      27.136       0.134  
5            48    0.488694      35.831       0.131  
6            48    0.362781      20.945       0.139  
7            48    0.477511      20.474       0.119  
8            48    0.040321       8.729       0.119  
9            48    0.040754       8.911       0.121

In [5]:
londres.tail()

LCLid         day  energy_median  energy_mean  energy_max  \
32987  MAC005416  2014-02-24         0.0660     0.111437       0.809   
32988  MAC005416  2014-02-25         0.0645     0.072521       0.238   
32989  MAC005416  2014-02-26         0.0675     0.107917       0.343   
32990  MAC005416  2014-02-27         0.0910     0.104250       0.296   
32991  MAC005416  2014-02-28         0.0630     0.063000       0.063   

       energy_count  energy_std  energy_sum  energy_min  
32987            48    0.126465       5.349       0.050  
32988            48    0.030756       3.481       0.051  
32989            48    0.080074       5.180       0.049  
32990            48    0.054844       5.004       0.047  
32991             1         NaN       0.063       0.063

Si queremos saber el nombre de las columnas o índices, podemos acceder a ellos:

In [6]:
londres.columns

Index(['LCLid', 'day', 'energy_median', 'energy_mean', 'energy_max',
       'energy_count', 'energy_std', 'energy_sum', 'energy_min'],
      dtype='object')

In [7]:
londres.columns[4]

'energy_max'

In [8]:
londres.index

RangeIndex(start=0, stop=32992, step=1)

Por útlimo, para tener un resumen muy bueno, es muy util el metodo ``describe``

In [9]:
londres.describe()

energy_median   energy_mean    energy_max  energy_count    energy_std  \
count   32992.000000  32992.000000  32992.000000  32992.000000  32865.000000   
mean        0.165632      0.223007      0.879596     47.776128      0.182819   
std         0.158396      0.178484      0.672407      3.035524      0.147394   
min         0.000000      0.000000      0.000000      1.000000      0.000000   
25%         0.052000      0.085432      0.348000     48.000000      0.069162   
50%         0.118500      0.176844      0.775000     48.000000      0.149607   
75%         0.225000      0.311276      1.248000     48.000000      0.264400   
max         1.608000      1.407458      6.897000     48.000000      1.104101   

         energy_sum    energy_min  
count  32992.000000  32992.000000  
mean      10.660024      0.064419  
std        8.565058      0.076632  
min        0.000000      0.000000  
25%        4.046000      0.016000  
50%        8.458000      0.035000  
75%       14.910500      0.093000  
max       67.558000      1.068000

### 10.3 Seleccionar datos

Si queremos trabajar o seleccionar solo con una parte del dataset, podemos usar **[]** después del nombre del dataset, e indicar qué columnas se desean seleccionar.

In [12]:
londres[['day', 'energy_max']].head()

day  energy_max
0  2011-12-14       1.674
1  2011-12-15       1.434
2  2011-12-16       2.034
3  2011-12-17       1.068
4  2011-12-18       2.218

Si, por el contrario se quiere seleccionar una parte de las filas, se indicará usando *slicing* como en los strings:

In [15]:
londres.iloc[0]

LCLid             MAC000113
day              2011-12-14
energy_median         1.091
energy_mean         1.06525
energy_max            1.674
energy_count             28
energy_std         0.305397
energy_sum           29.827
energy_min            0.262
Name: 0, dtype: object

#### loc vs iloc

Para seleccionar partes más específicas (más de una columna, varias columnas y solo unas filas de éstas,...), se usará la indexación con el loc o iloc. La única diferencia entre los dos, es que el primero se basa en el *label* de las filas o columnas, mientras que iloc se basa en el *integer* de estas. Con un ejemplo se entiende mejor:

In [ ]:
londres.loc[3:5,['day', 'energy_max']]

In [ ]:
londres.iloc[3:5,1:3]

Se debe tener en cuenta que el índice no será siempre un número y puede tratar-se de otro parámetro. Por ejemplo, podemos poner el día como índice.

### 10.4 Filtrar datos

Otra manera de seleccionar solo una parte del dataset es aplicar *Boolena indexing*, que no es nada más que aplicar un filtro a los datos. Por ejemplo, vamos a seleccionar solo aquellos días en los que la energia máxima ha sido mayor a 1.

In [ ]:
londres['energy_max'] > 1

Como se puede observar ahora tenemos seleccionadas aquellas columnas que cumplen nuestra condición. ¿Y si lo ponemos directamente dentro del dataset cuando seleccionamos solo una part?

In [ ]:
londres[londres['energy_max']>1]

#### Ejercicio

Obtener un cuadro resumen de los parámetros para el contador con ID 'MAC000113'

In [ ]:
# vuestro codigo aquí!


### 10.5 Funciones básicas interesantes

Aquí poeis ver una lista de las funciones más interesantes a mirar en la documentación, con algun ejemplo:

| Función   | Quantitativas | Qualitativas|
|-----------|---------------|-------------|
| count()   |Número de observaciones non-null|  
| sum()     |Suma de valores|
| mean()    |Media de las observaciones            | 
| median()  |Mediana de las observaciones             |
| min()     |Valor mínimo|
| max()     |Valor máximo|
| std()     |Desviación standard|
| var()     | Varianza|
| value_counts()     | Valores unicos en cada columna - tabla de frec|
| nunique()          | Número de valores distintos|
| isnull()     | Máscara de valores null o Nan|

Practicad con alguno de estos parámetros

In [ ]:
# vuestro código aquí!

# Valor total de energia para el Id 'MAC000113'

# ¿Cuántos contadores distintos hay?

# ¿Cómo se llaman?

# ¿Cuál fue el valor máximo de energía en dia 15 de Mayo de 2013?

### 10.6 Manipular dataframes

Se puede aplicar una función a alguna columna usando el método *apply* y recordando las fuciones lambda!

In [ ]:
londres['energy_var'] = londres['energy_std'].apply(lambda x: x**2)
londres.head()

#### Ejercicio

Crea una columna nueva llamada 'ID' con solo las últimas 6 cifras de 'LCLid'

In [ ]:
# vuestro código aquí!

### 10.7 Ordenar

Si se desea ordenar valores, se puede usar ``sort_values``. Con un ejemplo se entenderá mejor:


In [ ]:
londres.sort_values(by='energy_max', ascending= False, inplace=True)
londres.head()

### 10.8 Agrupar

Por último, una función muy útil es la de agrupar (group_by).

Por ejemplo, imagina que quieres agrupar todos los datos por número de contador y sacar algun parámetro de este contador (máximo, mínimo, etc.).

El resultado de esta operación, será otro dataframe con los datos agrupados respeto a una variable. Además también se deberá indicar qué función de *agregación* se desea usar (max, min, mean, count, sum, etc.).

Vamos a implementar el ejemplo comentado!

In [1]:
contadores = londres[['LCLid','energy_max', 'energy_min', 'energy_sum']].groupby('LCLid').mean()
contadores

NameError: name 'londres' is not defined

### 10.9 Merge

Muchas veces tendremos la información deseada en distintas bases de datos, archivos o datasets. Pandas nos ofrece la opción de juntar y fusionar datasets con *merge*.

Veamos un ejemplo con el dataset *londres*. 

Si cargaos otro dataset con la información de cada casa obtenemos:

In [9]:
import pandas as pd

londres = pd.read_csv('Data/block_13_diario.csv')
londres.head(5)

LCLid         day  energy_median  energy_mean  energy_max  \
0  MAC000113  2011-12-14         1.0910     1.065250       1.674   
1  MAC000113  2011-12-15         0.3710     0.546521       1.434   
2  MAC000113  2011-12-16         0.3205     0.534708       2.034   
3  MAC000113  2011-12-17         0.2570     0.421500       1.068   
4  MAC000113  2011-12-18         0.3985     0.565333       2.218   

   energy_count  energy_std  energy_sum  energy_min  
0            28    0.305397      29.827       0.262  
1            48    0.410120      26.233       0.119  
2            48    0.464431      25.666       0.129  
3            48    0.274387      20.232       0.115  
4            48    0.450712      27.136       0.134

In [14]:
casas = pd.read_csv('Data/informations_households.csv')
casas.head(5)

LCLid stdorToU    Acorn Acorn_grouped     file
0  MAC005492      ToU   ACORN-        ACORN-  block_0
1  MAC001074      ToU   ACORN-        ACORN-  block_0
2  MAC000002      Std  ACORN-A      Affluent  block_0
3  MAC003613      Std  ACORN-A      Affluent  block_0
4  MAC003597      Std  ACORN-A      Affluent  block_0

Podemos ver como por cada contador, nos da información de dónde podemos encontrar más parámetros de aquel contador y los agrupa por unos rupos llamados Acron. Para saber más sobre estos, ir directamente al repositorio de este dataset e inspeccionar con detalle.

De todos modos, si queremos juntar toda la información en un mismo dataset, tendríamos que hacer un merge:

<img src="Figures/Panda Data Wrangling Cheat Sheet 2.jpeg" alt="Drawing" style="width: 500px;"/>


Donde los modos *how* posibles se refieren a:

<img src="Figures/merge_options.png" alt="Drawing" style="width: 500px;"/>


In [18]:
combinado = pd.merge(londres, casas, how='left', on='LCLid')
combinado.head(5)

LCLid         day  energy_median  energy_mean  energy_max  \
0  MAC000113  2011-12-14         1.0910     1.065250       1.674   
1  MAC000113  2011-12-15         0.3710     0.546521       1.434   
2  MAC000113  2011-12-16         0.3205     0.534708       2.034   
3  MAC000113  2011-12-17         0.2570     0.421500       1.068   
4  MAC000113  2011-12-18         0.3985     0.565333       2.218   

   energy_count  energy_std  energy_sum  energy_min stdorToU    Acorn  \
0            28    0.305397      29.827       0.262      Std  ACORN-E   
1            48    0.410120      26.233       0.119      Std  ACORN-E   
2            48    0.464431      25.666       0.129      Std  ACORN-E   
3            48    0.274387      20.232       0.115      Std  ACORN-E   
4            48    0.450712      27.136       0.134      Std  ACORN-E   

  Acorn_grouped      file  
0      Affluent  block_13  
1      Affluent  block_13  
2      Affluent  block_13  
3      Affluent  block_13  
4      Affluent  block_13

### 10.10 Otros

Hay muchas más funciones que se pueden usar con Pandas. Os animamos a mirar en su documentación, foros, cheat sheets, cursos,...

https://pandas.pydata.org/pandas-docs/stable/index.html

## 11. Creación de gáficos

En Python existen diversas librerías para crear gráficos como Bokeh, altair, Pandas o Matplotlib

<img src="Figures/four_logos.png" alt="Drawing" style="width: 800px;"/>

En este curso aprenderemos las funciones básicas de matplotlib, ya que las otras librerías siempre estan basadas en esta y nos ayudará a luego saber usarlas también.

### 11.1 Matplotlib.pyplot

El módulo que se usa es el pyplot de matplotlib. Toda su documentación se puede encontrar aquí:

https://matplotlib.org/api/pyplot_api.html#module-matplotlib.pyplot

Vaamos a importarla!

In [ ]:
import matplotlib.pyplot as plt

Las dos instancias más importntes en pyplot son
* figure: en matplotlib.figure.Figure, una *figure* ser dónde iran los gráficos. Es el lienzo donde puede haber uno o varios gráficos.
* axes: eb matplotlib.axes.Axes, econtramos dode iran los gráficos en sí

Vamos a crear un primer gráfico de prueba:

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

plt.plot([1,2,3,4],[1,4,8,16])
plt.show()

En este caso hemos creado un gráfico usando *plot*, donde el primer argumento es el eje de las x y el segundo el de las y (deben ser de la smismas dimensiones). 

Podemos añadir otra información como título y nombre de los ejes:

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

plt.plot([1,2,3,4],[1,4,8,16])
plt.title('Nuestro primer plot!')
plt.xlabel('eje X')
plt.ylabel('eje Y')
plt.show()

### 11.2 Crear ventanas

Si se quiere modificar el tamaño del gráfico, es dóde tendremos que usar el término *figure* comentdo anteriormente. Recordamos que *figure* era el lienzo donde poníamos uno o más gráficos. En este caso, solo ponemos uno.

Aquí podemos ver como se puede configurar una *figure* con unas medidas predeterminadas:

In [ ]:
plt.figure(figsize=(20,5))
plt.plot([1,2,3,4],[1,4,8,16])
plt.show()

Para cada gráfico también se puede determinar el estilo del gráfico con un tercer argumento:

In [ ]:
plt.plot([1,2,3,4],[1,4,8,16],'go')
plt.show()

Y podemos poner más de un gráfico en el mismo plot, indicando en la leyenda que es cada cosa:

In [ ]:
plt.plot([1,2,3,4],[1,4,8,16],'g-',label='primero')
plt.plot([1,2,3,4],[1,6,12,6],'r-', label='segundo')
plt.title('Ahora son dos!')
plt.xlabel('eje X')
plt.ylabel('eje Y')
plt.legend()
plt.show()

### 11.3 Subplots

Para poner distintos gráficos con distintos ejes se puede usar Subplots. El métdodo `subplot` tiene 3 argumentos: número de filas, número de columnas e índice. Veamos con un ejemplo como se usa:

In [ ]:
plt.figure(figsize=(15,2))

plt.subplot(1,2,1)
plt.plot([1,2,3,4],[1,4,9,16],'go')
plt.title('Izquierda')
plt.xlabel('un eje X')
plt.ylabel('un eje Y')

plt.subplot(1,2,2)
plt.plot([1,2,3,4],[1,6,12,6],'r-')
plt.title('Derecha')
plt.xlabel('otro eje X')
plt.ylabel('otro eje Y')

O, por ejemplo:

In [ ]:
plt.figure(figsize=(10,10))

plt.subplot(2,1,1)
plt.plot([1,2,3,4],[1,4,9,16],'go')
plt.title('Arriba')
plt.xlabel('un eje X')
plt.ylabel('un eje Y')

plt.subplot(2,1,2)
plt.plot([1,2,3,4],[1,6,12,6],'r-')
plt.title('Abajo')
plt.xlabel('otro eje X')
plt.ylabel('otro eje Y')

### 11.4 Gráfico de barras

Para hacer un gráfico de barras en vez de `plot`usaremos `bar`, siguendo una estructura muy parecida.

Recordemos el dataset de antes:

In [ ]:
contadores = londres[['LCLid','energy_max', 'energy_min', 'energy_sum']].groupby('LCLid').mean()
contadores.head(10)

Podemo hacer una gràfica de barras:

In [ ]:
divisiones=contadores.index[:10]
marcas=contadores['energy_max'][:10]

plt.figure(figsize=(20,10))
plt.bar(divisiones, marcas, color = 'blue')
plt.xlabel('contador')
plt.ylabel('Energía máxima')
plt.xticks(rotation=45, ha='right')

O en barras horizontales, usando `barh`:

In [ ]:
divisiones=contadores.index[:10]
marcas=contadores['energy_max'][:10]

plt.figure(figsize=(20,10))
plt.barh(divisiones, marcas, color = 'blue')
plt.ylabel('Contador')
plt.xlabel('Energía máxima')

### 11.5 Histograma

Veamos con un ejemplo como hacer un histograma:

In [ ]:
plt.hist(londres[londres['LCLid']=='MAC000113']['energy_max'],8)

#### Ejercicio

¿Qué muestra el gráfico anterior? ¿Cómo lo podemos mejorar?

In [ ]:
#Vuestro código!